fix: custom widgets ios (#617)
ZupIT/beagle-docs: content/pt/resources/customization/beagle-for-ios/custom-widget/serverdrivencomponent-custom-widget.md

---

## Introdução

Nesse exemplo vamos ensinar a criar componentes que recebem um `ServerDrivenComponent`.

**`ServerDrivenComponent:`** São todos componentes default do beagle e componentes registrados no beagle.

{{% alert color="info" %}}
Para saber mais sobre como criar um componente veja, [**Widget Customizado Simples**]({{< ref path="/resources/customization/beagle-for-ios/custom-widget/simple-custom-widget" lang="pt" >}}).
{{% /alert %}}

### Passo 1: Criar o componente Customizado.

Abaixo temos a definição da classe do componente customizado ContainerTitle que recebe os parâmetros title e child.

**Parâmetros:**

**`title`**: Parâmetro do tipo `String` que representa o titulo do componente.

**`child`**: Parâmetro do tipo `UIView` que representa o view que o componente encapsula.

```swift
import Foundation
import UIKit

class ContainerTitle: UIView {
    
    // Class initialization.
    public init(title: String, child: UIView) {
        super.init(frame: .zero)
        setupView(child: child, title: title)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    

>     // Method to add component to hierarchy and position to superview.

Reviewer: ```suggestion
    // Adds the component to the hierarchy and the position to the Superview.
```

    private func setupView(child: UIView, title: String) {
        addSubview(label)
        label.text = title
        label.topAnchor.constraint(equalTo: topAnchor).isActive = true
        label.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10).isActive = true
        label.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -10).isActive = true
        
        addSubview(child)
        child.topAnchor.constraint(equalTo: label.bottomAnchor).isActive = true
        child.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10).isActive = true
        child.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -10).isActive = true
        child.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -10).isActive = true
        child.backgroundColor = .systemOrange
    }

>     // Component `UILabel` created.

Reviewer: ```suggestion
    // Creates component 'UILabel'
```

    private lazy var label: UILabel = {
        let label = UILabel()
        label.font = .systemFont(ofSize: 20, weight: .bold)
        label.backgroundColor = .orange
        label.textAlignment = .center
        label.textColor = .white
        label.translatesAutoresizingMaskIntoConstraints = false
        return label
    }()
}
```

### Passo 2: Criar o Widget.

> Agora temos o componente `ContainerTitle`, para trasformar para um componente Beagle temos adotar o protocolo `Widget`, que é um protocolo que conforma com `Decodable` e é responsável por decodificar as propriedades que seu widget expõem ao backend.

Reviewer: ```suggestion
Para transformar o componente `ContainerTitle` para um componente Beagle, temos que adotar o protocolo `Widget`, que é um protocolo compatível com `Decodable`, e é responsável por decodificar as propriedades que seu widget expõe no backend.
```

> ### Passo 2: Criar o Widget.
> 
> Agora temos o componente `ContainerTitle`, para trasformar para um componente Beagle temos adotar o protocolo `Widget`, que é um protocolo que conforma com `Decodable` e é responsável por decodificar as propriedades que seu widget expõem ao backend.
> 
> Crie uma struct ContainerTitleWidget adotando protocolo `Widget`, a interface widget irá adicionar a property de **widgetProperties** e o método **toView**.
> 
>  * **widgetProperties:** A propriedade de aplicar estilo, id e acessibilidade.
> 
>  * **toView:** Método para retornar a view do componente criado.
> 
> Temos a estrutura da struct `ContainerTitleWidget` com os parâmetros `title`, `child` e `widgetProperties` e o componente `ContainerTitle` criado no método `toView`.
> 
> Para exibir o componente recebido utizamos a `BeagleView`, logo a mesma retorna uma `UIView`. Para usar no meio do autoLayout é preciso desabilitar o `translatesAutoresizingMaskIntoConstraints` da view.

Reviewer 2: ```suggestion
### Passo 2: Criar o Widget

Agora é o componente `ContainerTitle`, para trasformá-lo em um componente Beagle, adote o protocolo `Widget`, que é compatível com o `Decodable` e é responsável por decodificar as propriedades que seu widget expõem no backend.

1. Crie uma estrutura **ContainerTitleWidget** adotando protocolo `Widget`. A interface widget irá adicionar a property de **widgetProperties** e o método **toView**.

 * **widgetProperties:** A propriedade de aplicar estilo, id e acessibilidade.

 * **toView:** Método para retornar a view do componente criado.

2. A estrutura do `ContainerTitleWidget` com os parâmetros `title`, `child` e `widgetProperties` e o componente `ContainerTitle` criado no método `toView`.

3. Para exibir o componente recebido, utilize a `BeagleView`, logo a mesma retorna uma `UIView`. Para usar no meio do autoLayout é preciso desabilitar o `translatesAutoresizingMaskIntoConstraints` da view.
```

```swift
import Foundation
import UIKit
import Beagle

struct ContainerTitleWidget: Widget {
    
    // Class parameter.
    let title: String
    var child: ServerDrivenComponent
    var widgetProperties: WidgetProperties
    
    // toView method of interface the  widget.
    func toView(renderer: BeagleRenderer) -> UIView {
        
        // BeagleView declaration.
        let child = BeagleView(child)
        child.translatesAutoresizingMaskIntoConstraints = false
        
         // Custom component declaration.
        let containerTitle = ContainerTitle(title: title, child: child)
        
    }
}
```

> 
> Temos que criar a parte de inicialização e decodificação do componente, tem duas maneiras possíveis usando o `sourcery` gerador de código para a linguagem Swift, ou fazendo manualmente.
> 
> **Sourcery:**
> 
> Para usar o `Sourcery` pode encontrar nesse [**link**]({{< ref path="/resources/customization/beagle-for-ios/sourcery" lang="pt" >}})
> 
> **Manual:**
> 
> Para fazer manual tem que criar o init e a decodificação dos parametros `title` e `widgetProperties` da struct `BoxWidget`.
> O widgetProperties tem sua propria parte decodificação, entao é preciso apenas passar o decoder para o objeto `WidgetProperties`.

Reviewer 2: ```suggestion

Crie a parte de inicialização e decodificação do componente e você pode fazer isso de duas formas: 
1. Usando o `sourcery` gerador de código para a linguagem Swift.

2. Você pode fazer manualmente. 

**Sourcery:**

Para saber mais sobre como usar o `Sourcery`, acesse a seção [**Sourcery do Beagle**]({{< ref path="/resources/customization/beagle-for-ios/sourcery" lang="pt" >}})

**Manual:**

Manualmente, crie o init e a decodificação dos parâmetros **`title`** e **`widgetProperties`** da struct `BoxWidget`.
O widgetProperties tem a sua própria parte decodificação, então é preciso apenas passar o decoder para o objeto `WidgetProperties`.
```

```swift

    // Initialization part of the class.
    init(
        title: String,
        child: ServerDrivenComponent,
        widgetProperties: WidgetProperties = WidgetProperties()
    ) {
        self.title = title
        self.child = child
        self.widgetProperties = widgetProperties
    }
    
    // Enum with parameters for decoding.
    enum CodingKeys: String, CodingKey {
        case title
        case child
    }

    // Initialization for decoding
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        title = try container.decode(String.self, forKey: .title)
        child = try container.decode(forKey: .child)
        widgetProperties = try WidgetProperties(from: decoder)
    }
```

> Para integrar o componente ao beagle é preciso utilizar o `sizeThatFits` ou `AutoLayoutWrapper`. 

Reviewer 2: ```suggestion
Integre o componente ao Beagle utilizando o `sizeThatFits` ou `AutoLayoutWrapper`. 
```

{{< tabs id="T0" >}}
{{% tab name="AutoLayoutWrapper" %}}

### AutoLayoutWrapper

**`AutoLayoutWrapper:`** O objeto calcula o tamanho levando em consideração as contraints do componente.
Para isso primeiro é preciso desabilitar o `translatesAutoresizingMaskIntoConstraints` da view do componente, e depois adicionar a view do componente dentro do `AutoLayoutWrapper`.

> Fazendo as configurações com o `AutoLayoutWrapper`.

Reviewer: Espaço a+ antes do Desabilite...

    private func setupView(child: UIView, title: String) {
        addSubview(label)
        label.text = title
        label.topAnchor.constraint(equalTo: topAnchor).isActive = true
        label.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10).isActive = true
        label.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -10).isActive = true
        
        addSubview(child)
        child.topAnchor.constraint(equalTo: label.bottomAnchor).isActive = true
        child.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10).isActive = true
        child.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -10).isActive = true
        child.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -10).isActive = true
        child.backgroundColor = .systemOrange
    }

    // Component `UILabel` created.
    private lazy var label: UILabel = {
        let label = UILabel()
        label.font = .systemFont(ofSize: 20, weight: .bold)
        label.backgroundColor = .orange
        label.textAlignment = .center
        label.textColor = .white
        label.translatesAutoresizingMaskIntoConstraints = false
        return label
    }()
}
```

> Classe completa do Widget.

Reviewer 2: ```suggestion
Veja a classe completa do Widget:
```

    init(
        title: String,
        child: ServerDrivenComponent,
        widgetProperties: WidgetProperties = WidgetProperties()
    ) {
        self.title = title
        self.child = child
        self.widgetProperties = widgetProperties
    }
    
    // Enum with parameters for decoding.
    enum CodingKeys: String, CodingKey {
        case title
        case child
    }

    // Initialization for decoding
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        title = try container.decode(String.self, forKey: .title)
        child = try container.decode(forKey: .child)
        widgetProperties = try WidgetProperties(from: decoder)
    }
    
    // toView method of interface the  widget.
    func toView(renderer: BeagleRenderer) -> UIView {

        // BeagleView declaration.
        let child = BeagleView(child)
        child.translatesAutoresizingMaskIntoConstraints = false
        
        // Custom component declaration.
        let containerTitle = ContainerTitle(title: title, child: child)
        
        return beagleWcontainerTitlerapper
    }
}
```

{{% /tab %}}
  {{< /tabs >}}

> ### Passo 3: Registrar o Widget.
> 
> É obrigatório **registrá-lo no Beagle.** Dentro do arquivo de configuração do beagle utilize o `dependencies` para registar. 
> 
> {{% alert color="info" %}}
> Para saber mais sobre o dependencies. [**Beagle Dependencies**]({{< ref path="/resources/customization/beagle-for-ios/beagles-dependencies" lang="pt" >}}).
> {{% /alert %}}
> 
> O método `register` possui dois construtores, o primeiro passando apenas o `component` e segundo recebendo o `component` e `named`.
> 
> * **component:** Passa a classe do componente.
> 
> * **named:** Parâmetro para setar o nome do componente. Não é obrigatório passar. Um caso é quando o nome do componente é registrado diferente com que você criou no backend. Ele será usado na deserializações para encontrar seu componente.
> 
> **Maneiras de Registrar**
> ```swift
> // 1
> dependencies.decoder.register(component: ContainerTitleWidget.self)
> 
> // 2
> dependencies.decoder.register(component: ContainerTitleWidget.self, named: "ContainerTitleWidgetComponent")
> ```
> 
> Após registrar o seu componente de customização, você pode usá-lo  via server-driven.

Reviewer 2: ```suggestion
### Passo 3: Registrar o Widget

É obrigatório **registrar o Widget no Beagle.** Dentro do arquivo de configuração, utilize o **`dependencies`** para registrar. 

{{% alert color="info" %}}
Para saber mais sobre o dependencies, veja a seção [**Beagle Dependencies**]({{< ref path="/resources/customization/beagle-for-ios/beagles-dependencies" lang="pt" >}}).
{{% /alert %}}

O método `register` possui dois construtores:
1: Passa apenas o `component`;
2:  Recebe o `component` e  o `named`.

* 1. **component:** Passa a classe do componente.

* 2. **named:** Parâmetro para setar o nome do componente. Não é obrigatório passar, por exemplo, quando o nome do componente é registrado diferente daquele que você criou no backend, ele será usado na deserializações para encontrar seu componente.

**Maneiras de Registrar**
```swift
// 1
dependencies.decoder.register(component: ContainerTitleWidget.self)

// 2
dependencies.decoder.register(component: ContainerTitleWidget.self, named: "ContainerTitleWidgetComponent")
```

Após registrar o seu componente de customização, você pode usá-lo  via server-driven.
```

---
title: Widget Customizado com ServerDrivenComponent
weight: 159
description: Criando um widget customizado com ServerDrivenComponent
---

## Como criar? 

Siga os passos abaixo para criar componentes que recebem um `ServerDrivenComponent`.

**`ServerDrivenComponent:`** São todos os componentes default e registrados no Beagle. 

{{% alert color="info" %}}

> Para saber mais sobre como criar um componente veja, [**Widget Customizado Simples**]({{< ref path="/resources/customization/beagle-for-ios/custom-widget/simple-custom-widget" lang="pt" >}}).

Reviewer: ```suggestion
Para saber mais sobre como criar um componente, veja [**Widget Customizado Simples**]({{< ref path="/resources/customization/beagle-for-ios/custom-widget/simple-custom-widget" lang="pt" >}}).
```

        addSubview(label)
        label.text = title
        label.topAnchor.constraint(equalTo: topAnchor).isActive = true
        label.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10).isActive = true
        label.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -10).isActive = true
        
        addSubview(child)
        child.topAnchor.constraint(equalTo: label.bottomAnchor).isActive = true
        child.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10).isActive = true
        child.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -10).isActive = true
        child.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -10).isActive = true
        child.backgroundColor = .systemOrange
    }

    // Creates component 'UILabel'
    private lazy var label: UILabel = {
        let label = UILabel()
        label.font = .systemFont(ofSize: 20, weight: .bold)
        label.backgroundColor = .orange
        label.textAlignment = .center
        label.textColor = .white
        label.translatesAutoresizingMaskIntoConstraints = false
        return label
    }()
}
```

### Passo 2: Criar o Widget

> Agora é o componente `ContainerTitle`, para trasformá-lo em um componente Beagle, adote o protocolo `Widget`, que é compatível com o `Decodable` e é responsável por decodificar as propriedades que seu widget expõem no backend.

Reviewer: ```suggestion
Agora falaremos do componente `ContainerTitle`. Para transformá-lo em um componente Beagle, adote o protocolo `Widget`, que é compatível com o `Decodable` e é responsável por decodificar as propriedades que seu widget expõe no backend.
```

1. Crie uma estrutura **ContainerTitleWidget** adotando protocolo `Widget`. A interface widget irá adicionar a property de **widgetProperties** e o método **toView**.

 * **widgetProperties:** A propriedade de aplicar estilo, id e acessibilidade.

 * **toView:** Método para retornar a view do componente criado.

2. A estrutura do `ContainerTitleWidget` com os parâmetros `title`, `child` e `widgetProperties` e o componente `ContainerTitle` criado no método `toView`.

> 3. Para exibir o componente recebido, utilize a `BeagleView`, logo a mesma retorna uma `UIView`. Para usar no meio do autoLayout é preciso desabilitar o `translatesAutoresizingMaskIntoConstraints` da view.

Reviewer: ```suggestion
3. Para exibir o componente recebido, utilize a `BeagleView`, já que a mesma retorna uma `UIView`. Para usar no meio do autoLayout é preciso desabilitar o `translatesAutoresizingMaskIntoConstraints` da view.
```

Reviewer: Verificar se esta sugestão muda a regra de negócio pretendida

```swift
import Foundation
import UIKit
import Beagle

struct ContainerTitleWidget: Widget {
    
    // Class parameter.
    let title: String
    var child: ServerDrivenComponent
    var widgetProperties: WidgetProperties
    
    // toView method of interface the  widget.
    func toView(renderer: BeagleRenderer) -> UIView {
        
        // BeagleView declaration.
        let child = BeagleView(child)
        child.translatesAutoresizingMaskIntoConstraints = false
        
         // Custom component declaration.
        let containerTitle = ContainerTitle(title: title, child: child)
        
    }
}
```

> Crie a parte de inicialização e decodificação do componente e você pode fazer isso de duas formas: 

Reviewer: ```suggestion
Crie a parte de inicialização e decodificação do componente. É possível ser feito de duas formas: 
```

1. Usando o `sourcery` gerador de código para a linguagem Swift.

> 2. Você pode fazer manualmente. 

Reviewer: ```suggestion
2. Manualmente. 
```

**Sourcery:**

Para saber mais sobre como usar o `Sourcery`, acesse a seção [**Sourcery do Beagle**]({{< ref path="/resources/customization/beagle-for-ios/sourcery" lang="pt" >}})

**Manual:**

Manualmente, crie o init e a decodificação dos parâmetros **`title`** e **`widgetProperties`** da struct `BoxWidget`.

> O widgetProperties tem a sua própria parte decodificação, então é preciso apenas passar o decoder para o objeto `WidgetProperties`.

Reviewer: O que você acha de repensarmos a palavra "decoder"? Será que o leitor vai entender do que se trata?

        self.widgetProperties = widgetProperties
    }
    
    // Enum with parameters for decoding.
    enum CodingKeys: String, CodingKey {
        case title
        case child
    }

    // Initialization for decoding
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        title = try container.decode(String.self, forKey: .title)
        child = try container.decode(forKey: .child)
        widgetProperties = try WidgetProperties(from: decoder)
    }
    
    // toView method of interface the  widget.
    func toView(renderer: BeagleRenderer) -> UIView {

        // BeagleView declaration.
        let child = BeagleView(child)
        child.translatesAutoresizingMaskIntoConstraints = false
        
        // Custom component declaration.
        let containerTitle = ContainerTitle(title: title, child: child)
        
        // Setting the beagle wrapper.
        containerTitle.translatesAutoresizingMaskIntoConstraints = false
        let beagleWrapper = AutoLayoutWrapper(view: containerTitle)
        
        return beagleWrapper
    }
}
```

  {{% /tab %}}

{{% tab name="SizeThatFits" %}}

### SizeThatFits

> **`sizeThatFits:`** Método para implementar sua lógica de tamanho, usado na classe do componente customizado.

Reviewer: ```suggestion
**`sizeThatFits:`** Método para implementar a lógica de tamanho. Este método é usado na classe do componente customizado.
```

```swift
override func sizeThatFits(_ size: CGSize) -> CGSize {
    systemLayoutSizeFitting(size)
}
```

A classe do componente customizado com `sizeThatFits` implementado.

```swift
import Foundation
import UIKit

class ContainerTitle: UIView {
    

>     // Initialization part of the class.

Reviewer: ```suggestion
    // Class initialization
```

    public init(title: String, child: UIView) {
        super.init(frame: .zero)
        setupView(child: child, title: title)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    

>     // Implementation sizeThatFits

Reviewer: Penso que este comentário não agrega muito, já que o método está bem claro

Author: Vou remover os comentários

    override func sizeThatFits(_ size: CGSize) -> CGSize {
        systemLayoutSizeFitting(size)
    }

>     // Method to add component to hierarchy and pass position.

Reviewer: ```suggestion
    // Adds component to hierarchy and configures position
```

    private func setupView(child: UIView, title: String) {
        addSubview(label)
        label.text = title
        label.topAnchor.constraint(equalTo: topAnchor).isActive = true
        label.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10).isActive = true
        label.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -10).isActive = true
        
        addSubview(child)
        child.topAnchor.constraint(equalTo: label.bottomAnchor).isActive = true
        child.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10).isActive = true
        child.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -10).isActive = true
        child.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -10).isActive = true
        child.backgroundColor = .systemOrange
    }

>     // Component `UILabel` created.

Reviewer: ```suggestion
    // Component `UILabel`
```

        child: ServerDrivenComponent,
        widgetProperties: WidgetProperties = WidgetProperties()
    ) {
        self.title = title
        self.child = child
        self.widgetProperties = widgetProperties
    }
    
    // Enum with parameters for decoding.
    enum CodingKeys: String, CodingKey {
        case title
        case child
    }

    // Initialization for decoding
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        title = try container.decode(String.self, forKey: .title)
        child = try container.decode(forKey: .child)
        widgetProperties = try WidgetProperties(from: decoder)
    }
    
    // toView method of interface the  widget.
    func toView(renderer: BeagleRenderer) -> UIView {

        // BeagleView declaration.
        let child = BeagleView(child)
        child.translatesAutoresizingMaskIntoConstraints = false
        
        // Custom component declaration.
        let containerTitle = ContainerTitle(title: title, child: child)
        
        return beagleWcontainerTitlerapper
    }
}
```

{{% /tab %}}
  {{< /tabs >}}

### Passo 3: Registrar o Widget.

> É obrigatório **registrá-lo no Beagle.** Dentro do arquivo de configuração do beagle utilize o `dependencies` para registar. 

Reviewer: ```suggestion
É obrigatório **registrá-lo no Beagle.** Dentro do arquivo de configuração do Beagle, utilize o `dependencies` para registar. 
```

Reviewer: Está claro onde fica este arquivo?

{{% alert color="info" %}}
Para saber mais sobre o dependencies. [**Beagle Dependencies**]({{< ref path="/resources/customization/beagle-for-ios/beagles-dependencies" lang="pt" >}}).
{{% /alert %}}

O método `register` possui dois construtores, o primeiro passando apenas o `component` e segundo recebendo o `component` e `named`.

> * **component:** Passa a classe do componente.

Reviewer: ```suggestion
* **component:** Classe do componente.
```

> * **named:** Parâmetro para setar o nome do componente. Não é obrigatório passar. Um caso é quando o nome do componente é registrado diferente com que você criou no backend. Ele será usado na deserializações para encontrar seu componente.

Reviewer: ```suggestion
* **named:** Nome do componente. Não é obrigatório informar-lo. Utiliza-se por exemplo quando o nome do componente é registrado diferentemente com que você criou no backend. Ele será usado na deserializações para encontrar seu componente.
```

---
title: Widget Customizado com ServerDrivenComponent
weight: 159
description: Criando um widget customizado com ServerDrivenComponent
---

## Como criar? 

Siga os passos abaixo para criar componentes que recebem um `ServerDrivenComponent`.

**`ServerDrivenComponent:`** São todos os componentes default e registrados no Beagle. 

{{% alert color="info" %}}
Para saber mais sobre como criar um componente, veja [**Widget Customizado Simples**]({{< ref path="/resources/customization/beagle-for-ios/custom-widget/simple-custom-widget" lang="pt" >}}).
{{% /alert %}}

> ### Passo 1: Criar o componente Customizado

Reviewer: ```suggestion
### Passo 1: Criar o componente customizado
```

        addSubview(label)
        label.text = title
        label.topAnchor.constraint(equalTo: topAnchor).isActive = true
        label.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10).isActive = true
        label.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -10).isActive = true
        
        addSubview(child)
        child.topAnchor.constraint(equalTo: label.bottomAnchor).isActive = true
        child.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10).isActive = true
        child.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -10).isActive = true
        child.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -10).isActive = true
        child.backgroundColor = .systemOrange
    }

    // Creates component 'UILabel'
    private lazy var label: UILabel = {
        let label = UILabel()
        label.font = .systemFont(ofSize: 20, weight: .bold)
        label.backgroundColor = .orange
        label.textAlignment = .center
        label.textColor = .white
        label.translatesAutoresizingMaskIntoConstraints = false
        return label
    }()
}
```

### Passo 2: Criar o Widget

> Agora falaremos do componente `ContainerTitle`. Para transformá-lo em um componente Beagle, adote o protocolo `Widget`, que é compatível com o `Decodable` e é responsável por decodificar as propriedades que seu widget expõe no backend.

Reviewer 2: ```suggestion
Para transformar o componente `ContainerTitle` em um componente Beagle, adote o protocolo `Widget`. Ele é compatível com o `Decodable` e responsável por decodificar as propriedades que seu widget expõe no backend.
```

> 1. Crie uma estrutura **ContainerTitleWidget** adotando protocolo `Widget`. A interface widget irá adicionar a property de **widgetProperties** e o método **toView**.
> 
>  * **widgetProperties:** Propriedade para aplicar estilo, id e acessibilidade.
> 
>  * **toView:** Método para retornar a view do componente criado.
> 
> 2. A estrutura do `ContainerTitleWidget` com os parâmetros `title`, `child` e `widgetProperties`, e o componente `ContainerTitle` criado no método `toView`.
> 
> 3. Para exibir o componente recebido, utilize a `BeagleView`, já que a mesma retorna uma `UIView`. Para usar no meio do autoLayout é preciso desabilitar o `translatesAutoresizingMaskIntoConstraints` da view.

Reviewer 2: ```suggestion
Para criar um Widget: 
1. Crie uma estrutura **ContainerTitleWidget** e adote o protocolo `Widget`. A interface widget irá adicionar a property de **widgetProperties** e o método **toView**.

 * **widgetProperties:** Propriedade para aplicar estilo, id e acessibilidade;

 * **toView:** Método para retornar a view do componente criado.

2. Adicione a estrutura do `ContainerTitleWidget` com os parâmetros `title`, `child` e `widgetProperties`, e o componente `ContainerTitle`, já criado no método `toView`.

3. Utilize a `BeagleView` para exibir o componente recebido, ela retorna uma `UIView`. 
Se você quiser usar essa classe no meio do autoLayout é preciso desabilitar o `translatesAutoresizingMaskIntoConstraints` da view.
```

Reviewer 2: - Eu adicionei algumas palavras que entendi que faltavam no texto, olhe se faz sentido ou se precisa alterar, please. ;)

```swift
import Foundation
import UIKit
import Beagle

struct ContainerTitleWidget: Widget {
    
    // Class parameter.
    let title: String
    var child: ServerDrivenComponent
    var widgetProperties: WidgetProperties
    
    // toView method of interface the  widget.
    func toView(renderer: BeagleRenderer) -> UIView {
        
        // BeagleView declaration.
        let child = BeagleView(child)
        child.translatesAutoresizingMaskIntoConstraints = false
        
         // Custom component declaration.
        let containerTitle = ContainerTitle(title: title, child: child)
        
    }
}
```

Crie a parte de inicialização e decodificação do componente. É possível ser feito de duas formas: 

> 1. Usando o `sourcery` gerador de código para a linguagem Swift.

Reviewer 3: ```suggestion
1. Usando o  gerador de código para a linguagem Swift: `sourcery`.
```

2. Manualmente. 

**Sourcery:**

Para saber mais sobre como usar o `Sourcery`, acesse a seção [**Sourcery do Beagle**]({{< ref path="/resources/customization/beagle-for-ios/sourcery" lang="pt" >}})

**Manual:**

Manualmente, crie o init e a decodificação dos parâmetros **`title`** e **`widgetProperties`** da struct `BoxWidget`.

> O widgetProperties tem a sua própria parte decodificação, então é preciso apenas passar o decoder para o objeto `WidgetProperties`.

Reviewer 3: Aqui você quis dizer "O widgetProperties tem a sua própria parte DE decodificação" ?

```swift

    // Initialization part of the class.
    init(
        title: String,
        child: ServerDrivenComponent,
        widgetProperties: WidgetProperties = WidgetProperties()
    ) {
        self.title = title
        self.child = child
        self.widgetProperties = widgetProperties
    }
    
    // Enum with parameters for decoding.
    enum CodingKeys: String, CodingKey {
        case title
        case child
    }

    // Initialization for decoding
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        title = try container.decode(String.self, forKey: .title)
        child = try container.decode(forKey: .child)
        widgetProperties = try WidgetProperties(from: decoder)
    }
```

Integre o componente ao Beagle utilizando o `sizeThatFits` ou `AutoLayoutWrapper`. 

{{< tabs id="T0" >}}
{{% tab name="AutoLayoutWrapper" %}}

### AutoLayoutWrapper

**`AutoLayoutWrapper:`** O objeto calcula o tamanho considerando as restrições do componente.
Para fazer isso:

> -  Desabilite o `translatesAutoresizingMaskIntoConstraints` da view do componente;

Reviewer: ```suggestion
- Desabilite o `translatesAutoresizingMaskIntoConstraints` da view do componente;
```

        self.widgetProperties = widgetProperties
    }
    
    // Enum with parameters for decoding.
    enum CodingKeys: String, CodingKey {
        case title
        case child
    }

    // Initialization for decoding
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        title = try container.decode(String.self, forKey: .title)
        child = try container.decode(forKey: .child)
        widgetProperties = try WidgetProperties(from: decoder)
    }
    
    // toView method of interface the  widget.
    func toView(renderer: BeagleRenderer) -> UIView {

        // BeagleView declaration.
        let child = BeagleView(child)
        child.translatesAutoresizingMaskIntoConstraints = false
        
        // Custom component declaration.
        let containerTitle = ContainerTitle(title: title, child: child)
        
        // Setting the beagle wrapper.
        containerTitle.translatesAutoresizingMaskIntoConstraints = false
        let beagleWrapper = AutoLayoutWrapper(view: containerTitle)
        
        return beagleWrapper
    }
}
```

  {{% /tab %}}

{{% tab name="SizeThatFits" %}}

### SizeThatFits

> **`sizeThatFits:`** Método para implementar a lógica de tamanho. Este método é usado na classe do componente customizado.

Reviewer 2: ```suggestion
**`sizeThatFits:`** É um método para implementar a lógica de tamanho. É usado na classe do componente customizado.
```

    init(
        title: String,
        child: ServerDrivenComponent,
        widgetProperties: WidgetProperties = WidgetProperties()
    ) {
        self.title = title
        self.child = child
        self.widgetProperties = widgetProperties
    }
    
    // Enum with parameters for decoding.
    enum CodingKeys: String, CodingKey {
        case title
        case child
    }

    // Initialization for decoding
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        title = try container.decode(String.self, forKey: .title)
        child = try container.decode(forKey: .child)
        widgetProperties = try WidgetProperties(from: decoder)
    }
    
    // toView method of interface the  widget.
    func toView(renderer: BeagleRenderer) -> UIView {

        // BeagleView declaration.
        let child = BeagleView(child)
        child.translatesAutoresizingMaskIntoConstraints = false
        
        // Custom component declaration.
        let containerTitle = ContainerTitle(title: title, child: child)
        
        return beagleWcontainerTitlerapper
    }
}
```

{{% /tab %}}
  {{< /tabs >}}

> ### Passo 3: Registrar o Widget.
> 
> É obrigatório **registrá-lo no Beagle.** Dentro do arquivo de configuração do Beagle, utilize o `dependencies` para registar. 
> 
> {{% alert color="info" %}}
> Para saber mais sobre o dependencies. [**Beagle Dependencies**]({{< ref path="/resources/customization/beagle-for-ios/beagles-dependencies" lang="pt" >}}).
> {{% /alert %}}
> 
> O método `register` possui dois construtores, o primeiro passando apenas o `component` e segundo recebendo o `component` e `named`.
> 
> * **component:** Classe do componente.
> 
> * **named:** Nome do componente. Não é obrigatório informar-lo. Utiliza-se por exemplo quando o nome do componente é registrado diferentemente com que você criou no backend. Ele será usado na deserializações para encontrar seu componente.

Reviewer 2: ```suggestion
### Passo 3: Registrar o Widget

É obrigatório **registrar o Widget no Beagle.** Para fazer isso, acesse o arquivo de configuração do Beagle e utilize o `dependencies` para registrar. 

{{% alert color="info" %}}
Para saber mais sobre o dependencies, veja a seção [**Beagle Dependencies**]({{< ref path="/resources/customization/beagle-for-ios/beagles-dependencies" lang="pt" >}}).
{{% /alert %}}

O método `register` possui dois construtores:
1. Passa apenas o `component`;
2. Recebe o `component` e `named`.

* **component:** A classe do componente.

* **named:** É o nome do componente e não é obrigatório informá-lo. Você pode usar quando você registrar um nome de componente diferente do que você criou no backend. 
Ele é usado na deserializações para encontrar o seu componente.
```

**Maneiras de Registrar**
```swift
// 1
dependencies.decoder.register(component: ContainerTitleWidget.self)

// 2
dependencies.decoder.register(component: ContainerTitleWidget.self, named: "ContainerTitleWidgetComponent")
```

Após registrar o seu componente de customização, você pode usá-lo  via server-driven.

> ### Passo 4: Exibir o Componente.
> 
> Você pode usar o seu componente declarativamente, passá-lo por uma instância até o `BeagleScreenViewController` ou chamá-lo via método `toView()` para apresentar o`UIView` que aparece dentro do seu próprio view controller.

Reviewer 2: ```suggestion
### Passo 4: Exibir o componente

Você pode usar o seu componente declarativamente e passá-lo por uma instância até o `BeagleScreenViewController`. A outra opção é chamá-lo via o método `toView()` para apresentar o`UIView` que aparece dentro do seu próprio view controller.
```

```swift
let beagleScreenViewController = Beagle.screen(
    .declarative(
        .init(child:
            ContainerTitleWidget(
                    title: "Title",
                    child: Container {
                        Text("Label 1")
                        Text("Label 2")
                    }
                )
        )
    )
)
```

{{% alert color="info" %}}
Para saber mais como exibir o Componente, veja a seção [**Como exibir uma tela**]({{< ref path="/get-started/using-beagle/ios" lang="pt" >}}).
{{% /alert %}}

> Veja abaixo um exemplo renderizado:

Reviewer 2: ```suggestion
Veja abaixo um exemplo de um componente renderizado:
```
